### PR TITLE
Added RollUpApply plan and pipe

### PR DESCRIPTION
**Pull request by systay**
RollUp is the inverse of the Unwind operator. For each
row passed in from the LHS, the whole RHS is executed.
For each row produced by the RHS, a single column
value is extracted and inserted into a collection.

It is used for sub queries that return collections,
such as pattern expressions (returns a collection of paths)
and pattern comprehension.

### CODE REVIEW

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Foldable.scala [hunk 1]
```diff
@@ -153,6 +156,19 @@ object Foldable {
...
+      Seq.empty
+    } else {
+      val that = remaining.pop()
+      val thisElement = that match {
+        case x: A => Seq(x)
+        case _ => Seq.empty
+      }
+
+      thisElement ++ findAllAcc(remaining ++= that.reverseChildren)
+    }
```
Comment (davidegrohmann) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Foldable.scala]: I think it is a good idea if we try to write these functions in a tailrec way

**Comment by systay** [comment]
Good point, @davidegrohmann. Think this version looks good? I wasn't able to make it `@tailrec`

**Comment by systay** [comment]
Just found a problem with this PR. Let's hold off merging it for a bit

**Comment by davidegrohmann** [comment]
@systay or:

```
def findByAllClass[A : Manifest]: Seq[A] =
      findAllAcc[A](mutable.ArrayStack(that), mutable.ArrayStack())


@tailrec
private def findAllAcc[A: ClassTag](remaining: scala.collection.mutable.ArrayStack[Any], partialResult: scala.collection.mutable.ArrayStack[A]): Seq[A] =
  if (remaining.isEmpty) {
    partialResult
  } else {
    val that = remaining.pop()
    that match {
      case x: A => partialResult.push(x)
      case _ =>
    }

    findAllAcc(remaining ++= that.reverseChildren, partialResult)
  }
```

**Comment by systay** [comment]
@davidegrohmann Do you think this non-recursive version is OK?

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Foldable.scala [hunk 2]
```diff
@@ -89,6 +89,9 @@ object Foldable {
 
     def findByClass[A : Manifest]: A =
       findAcc[A](mutable.ArrayStack(that))
+
+    def findByAllClass[A : Manifest]: Seq[A] =
```
Comment (pontusmelke) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Foldable.scala]: Can't we use ClassTag everywhere?

**Comment by davidegrohmann** [comment]
@systay it is fine, don't see though the point to get the `remaining` stack as a param rather than to create in the `findAllAcc` method.  Which leads to the fact that you don't need an "acc-method" anymore.

**Pull request merged**